File: Alcoholtracker/Views/Components/DurationChipRow.swift
import SwiftUI

// MARK: - DurationChipRow
//
// Picks how long a drink is consumed over ("verzögerter Start" / sipping). The
// value maps to Drink.drinkDurationMinutes, where 0 means "auto-estimate" and a
// positive value stretches the BAC absorption window (see BACCalculator
// .absorptionWindowMinutes), flattening and lowering the peak for slowly sipped
// drinks. Short durations may match the auto estimate and leave the curve
// unchanged; longer ones (a bottle of wine over two hours) move it noticeably.

struct DurationChipRow: View {
    @Binding var durationMinutes: Double
    let estimatedMinutes: Double?

    // (label, minutes). 0 = auto-estimate from category + volume.
    private let options: [(String, Double)] = [
        ("Auto", 0),
        ("30 min", 30),
        ("1 Std", 60),
        ("2 Std", 120),
        ("3 Std", 180),
    ]

    init(durationMinutes: Binding<Double>, estimatedMinutes: Double? = nil) {
        self._durationMinutes = durationMinutes
        self.estimatedMinutes = estimatedMinutes
    }

    private var isCustom: Bool {
        durationMinutes > 0 && !options.dropFirst().contains { Int($0.1) == Int(durationMinutes) }
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 10) {
            HStack(spacing: 8) {
                ForEach(options, id: \.1) { option in
                    let isSelected = Int(durationMinutes) == Int(option.1)
                    Button {
                        durationMinutes = option.1
                    } label: {
                        VStack(spacing: 2) {
                            Text(option.0)
                                .font(.appCaption)
                            if option.1 == 0, let estimatedMinutes {
                                Text(formatDuration(estimatedMinutes))
                                    .font(.appMicro)
                            }
                        }
                        .foregroundStyle(isSelected ? Color.appBackground : Color.appTextDim)
                        .frame(maxWidth: .infinity)
                        .padding(.vertical, option.1 == 0 && estimatedMinutes != nil ? 6 : 9)
                        .background(isSelected ? Color.appAccent : Color.appCard)
                        .clipShape(RoundedRectangle(cornerRadius: 10))
                        .overlay(
                            RoundedRectangle(cornerRadius: 10)
                                .strokeBorder(isSelected ? Color.appAccent : Color.appBorder, lineWidth: 0.5)
                        )
                    }
                    .buttonStyle(.pressable)
                    .animation(.appSnappy, value: isSelected)
                }
            }

            HStack(spacing: 10) {
                Button {
                    if durationMinutes <= 0 {
                        durationMinutes = estimatedMinutes ?? 30
                    }
                } label: {
                    Label("Eigene Dauer", systemImage: "slider.horizontal.3")
                        .font(.appCaptionBold)
                        .foregroundStyle(isCustom ? Color.appBackground : Color.appTextDim)
                        .frame(maxWidth: .infinity)
                        .padding(.vertical, 10)
                        .background(isCustom ? Color.appAccent : Color.appCard)
                        .clipShape(RoundedRectangle(cornerRadius: 10))
                        .overlay(
                            RoundedRectangle(cornerRadius: 10)
                                .strokeBorder(isCustom ? Color.appAccent : Color.appBorder, lineWidth: 0.5)
                        )
                }
                .buttonStyle(.pressable)

                if durationMinutes > 0 {
                    HStack(spacing: 6) {
                        Text(formatDuration(durationMinutes))
                            .font(.appCaptionBold)
                            .foregroundStyle(Color.appText)
                            .monospacedDigit()
                            .frame(minWidth: 56, alignment: .leading)
                        Stepper("", value: $durationMinutes, in: 1...240, step: 5)
                            .labelsHidden()
                    }
                    .padding(.horizontal, 12)
                    .padding(.vertical, 8)
                    .background(Color.appCard)
                    .clipShape(RoundedRectangle(cornerRadius: 10))
                    .overlay(
                        RoundedRectangle(cornerRadius: 10)
                            .strokeBorder(Color.appBorder, lineWidth: 0.5)
                    )
                }
            }
        }
    }

    private func formatDuration(_ minutes: Double) -> String {
        let rounded = max(1, Int(minutes.rounded()))
        if rounded < 60 { return "\(rounded) min" }
        let hours = rounded / 60
        let mins = rounded % 60
        return mins == 0 ? "\(hours) Std" : "\(hours)h \(mins)m"
    }
}

File: android/app/src/main/kotlin/de/tipau/promille/ui/components/DurationChipRow.kt
package de.tipau.promille.ui.components
import androidx.compose.material3.Icon
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*

import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import de.tipau.promille.appSpec
import de.tipau.promille.AppColors
import kotlin.math.roundToInt

/**
 * Port of DurationChipRow.swift.
 *
 * Picks how long a drink is consumed over ("verzögerter Start" / sipping).
 * 0 means "auto-estimate" and a positive value stretches the BAC absorption window.
 */
@Composable
fun DurationChipRow(
    durationMinutes: Double,
    onDurationMinutesChange: (Double) -> Unit,
    estimatedMinutes: Double? = null,
    modifier: Modifier = Modifier
) {
    // (label, minutes). 0 = auto-estimate from category + volume.
    val options = listOf(
        "Auto" to 0.0,
        "30 min" to 30.0,
        "1 Std" to 60.0,
        "2 Std" to 120.0,
        "3 Std" to 180.0
    )

    val isCustom = durationMinutes > 0 && options.drop(1).none { it.second.toInt() == durationMinutes.toInt() }

    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(10.dp)
    ) {
        // Preset duration chips
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            options.forEach { option ->
                val isSelected = durationMinutes.toInt() == option.second.toInt()
                val bgColor by animateColorAsState(
                    targetValue = if (isSelected) AppColors.accent else AppColors.card,
                    animationSpec = appSpec(tween(150)),
                    label = "durationChipBg"
                )
                val textColor by animateColorAsState(
                    targetValue = if (isSelected) AppColors.background else AppColors.textDim,
                    animationSpec = appSpec(tween(150)),
                    label = "durationChipText"
                )
                val borderColor = if (isSelected) AppColors.accent else AppColors.border

                Box(
                    modifier = Modifier
                        .weight(1f)
                        .clip(RoundedCornerShape(10.dp))
                        .background(bgColor)
                        .border(0.5.dp, borderColor, RoundedCornerShape(10.dp))
                        .clickable { onDurationMinutesChange(option.second) }
                        .padding(vertical = if (option.second == 0.0 && estimatedMinutes != null) 6.dp else 9.dp),
                    contentAlignment = Alignment.Center
                ) {
                    Column(
                        horizontalAlignment = Alignment.CenterHorizontally,
                        verticalArrangement = Arrangement.spacedBy(2.dp)
                    ) {
                        Text(
                            // iOS: .appCaption, no weight override (was
                            // 12sp Medium here).
                            text = option.first,
                            color = textColor,
                            style = de.tipau.promille.AppText.caption,
                            textAlign = TextAlign.Center
                        )
                        if (option.second == 0.0 && estimatedMinutes != null) {
                            Text(
                                text = formatDuration(estimatedMinutes),
                                color = textColor,
                                style = de.tipau.promille.AppText.micro,
                                textAlign = TextAlign.Center
                            )
                        }
                    }
                }
            }
        }

        // Custom duration button and stepper
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(10.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            val customBgColor by animateColorAsState(
                targetValue = if (isCustom) AppColors.accent else AppColors.card,
                animationSpec = appSpec(tween(150)),
                label = "customDurationBg"
            )
            val customTextColor by animateColorAsState(
                targetValue = if (isCustom) AppColors.background else AppColors.textDim,
                animationSpec = appSpec(tween(150)),
                label = "customDurationText"
            )
            val customBorderColor = if (isCustom) AppColors.accent else AppColors.border

            // "Eigene Dauer" button
            Box(
                modifier = Modifier
                    .weight(1f)
                    .clip(RoundedCornerShape(10.dp))
                    .background(customBgColor)
                    .border(0.5.dp, customBorderColor, RoundedCornerShape(10.dp))
                    .clickable {
                        if (durationMinutes <= 0) {
                            onDurationMinutesChange(estimatedMinutes ?: 30.0)
                        }
                    }
                    .padding(vertical = 10.dp, horizontal = 12.dp),
                contentAlignment = Alignment.Center
            ) {
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(6.dp)
                ) {
                    Icon(
                        imageVector = AppIcons.Sliders,
                        contentDescription = null,
                        tint = customTextColor,
                        modifier = Modifier.size(16.dp)
                    )
                    Text(
                        // iOS: .appCaptionBold (13sp SemiBold) - was 12sp Bold.
                        text = "Eigene Dauer",
                        color = customTextColor,
                        style = de.tipau.promille.AppText.captionBold
                    )
                }
            }

            // Stepper controls when duration > 0
            if (durationMinutes > 0) {
                Row(
                    modifier = Modifier
                        .clip(RoundedCornerShape(10.dp))
                        .background(AppColors.card)
                        .border(0.5.dp, AppColors.border, RoundedCornerShape(10.dp))
                        .padding(horizontal = 12.dp, vertical = 8.dp),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    Text(
                        // iOS: .appCaptionBold - was 12sp Bold.
                        text = formatDuration(durationMinutes),
                        color = AppColors.text,
                        style = de.tipau.promille.AppText.captionBold
                    )

                    // Minus button
                    Box(
                        modifier = Modifier
                            .size(24.dp)
                            .clip(RoundedCornerShape(6.dp))
                            .background(AppColors.background)
                            .border(0.5.dp, AppColors.border, RoundedCornerShape(6.dp))
                            .clickable {
                                val next = maxOf(1.0, durationMinutes - 5.0)
                                onDurationMinutesChange(next)
                            },
                        contentAlignment = Alignment.Center
                    ) {
                        Text(
                            text = "-",
                            color = AppColors.text,
                            fontSize = 14.sp,
                            fontWeight = FontWeight.Bold
                        )
                    }

                    // Plus button
                    Box(
                        modifier = Modifier
                            .size(24.dp)
                            .clip(RoundedCornerShape(6.dp))
                            .background(AppColors.background)
                            .border(0.5.dp, AppColors.border, RoundedCornerShape(6.dp))
                            .clickable {
                                val next = minOf(240.0, durationMinutes + 5.0)
                                onDurationMinutesChange(next)
                            },
                        contentAlignment = Alignment.Center
                    ) {
                        Text(
                            text = "+",
                            color = AppColors.text,
                            fontSize = 14.sp,
                            fontWeight = FontWeight.Bold
                        )
                    }
                }
            }
        }
    }
}

private fun formatDuration(minutes: Double): String {
    val rounded = maxOf(1, minutes.roundToInt())
    if (rounded < 60) return "$rounded min"
    val hours = rounded / 60
    val mins = rounded % 60
    return if (mins == 0) "$hours Std" else "${hours}h ${mins}m"
}
